smithy-jmespath/src/main/java/software/amazon/smithy/jmespath/evaluation/JmespathRuntime.java

```diff
@@ -0,0 +1,289 @@
+
+    /**
+     * If the given value is a STRING, return it as a String.
+     * Otherwise, throws a JmespathException of type INVALID_TYPE.
+     * <p>
+     * Note the distinction between this method and toString(),
+     * which can be called on any value and produces a JSON string.
+     */
+    String asString(T value);
+
+    ///////////////////////////////
+    // NUMBERs
+    ///////////////////////////////
+
+    T createNumber(Number value);
+
+    /**
+     * Returns the type of Number that asNumber() will produce for this value.
+     * Will be more efficient for some runtimes than checking the class of asNumber().
+     */
+    NumberType numberType(T value);
+
+    /**
+     * If the given value is a NUMBER, return it as a Number.
+     * Otherwise, throws a JmespathException of type INVALID_TYPE.
+     */
+    Number asNumber(T value);
+
+    ///////////////////////////////
+    // ARRAYs
+    ///////////////////////////////
+
+    ArrayBuilder<T> arrayBuilder();
+
+    interface ArrayBuilder<T> {
+
+        void add(T value);
+
+        void addAll(T array);
+
+        T build();
+    }
+
+    /**
+     * If the given value is an ARRAY, returns the element at the given index.
+     * Otherwise, throws a JmespathException of type INVALID_TYPE.
+     */
+    T element(T array, T index);
+
+    /**
+     * If the given value is an ARRAY, returns the specified slice.
+     * Otherwise, throws a JmespathException of type INVALID_TYPE.
+     * <p>
+     * The start and stop values will always be non-null and non-negative.
+     * Step will always be non-zero.
+     * If step is positive, start will be less than or equal to stop.
+     * If step is negative, start will be greater than or equal to stop.
+     */
+    default T slice(T array, Number startNumber, Number stopNumber, Number stepNumber) {
+        JmespathRuntime.ArrayBuilder<T> output = arrayBuilder();
+        int start = startNumber.intValue();
+        int stop = stopNumber.intValue();
+        int step = stepNumber.intValue();
+
+        if (start < stop) {
+            if (step > 0) {
+                for (int idx = start; idx < stop; idx += step) {
+                    output.add(element(array, createNumber(idx)));
+                }
+            }
+        } else {
+            if (step < 0) {
+                // List is iterating in reverse
+                for (int idx = start; idx > stop; idx += step) {
+                    output.add(element(array, createNumber(idx)));
+                }
+            }
+        }
+        return output.build();
+    }
+
+    ///////////////////////////////
+    // OBJECTs
+    ///////////////////////////////
+
+    ObjectBuilder<T> objectBuilder();
+
+    interface ObjectBuilder<T> {
+
+        void put(T key, T value);
+
+        void putAll(T object);
+
+        T build();
+    }
+
+    /**
+     * If the given value is an OBJECT, returns the value mapped to the given key.
+     * Otherwise, returns NULL.
```

Comment:
like a Java null? `NULL` reads like it's special. Is it the result of calling createNull on the Runtime?